distribution/push.go

```diff
@@ -127,6 +133,19 @@ func Push(ctx context.Context, ref reference.Named, imagePushConfig *ImagePushCo
 			continue
 		}
 
+		parsedURL, urlErr := url.Parse(endpoint.URL)
```

Comment:
@calavera: This is a fair point - quite possibly one I would raise myself if I reviewed similar code. I thought about this while working on the PR, but didn't immediately see a clean way to split the code out.

The new code here involves parsing the URL and also maintaining a map. The direct way to reuse it would be to provide a function that parses the URL, checks for it in the map, and returns the parsed representation, and also a function that adds the URL to the map if appropriate. Both functions would have to take a fair amount of state in arguments or a special structure, and it seemed like it would be ugly.

The `Push` and `Pull` functions are nearly identical, except one calls `newPuller`/`Pull` and the other calls `newPusher`/`Push` (and they have log messages tailored to the particular operation). Ideally `Push` and `Pull` would be combined into something more generic. The obstacle is that `Pusher` and `Puller` are different interfaces, so I don't see an obvious way to write a clean generic version.

I will experiment with this and get back to you.